analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs

```diff
@@ -43,11 +43,13 @@ protected override ProgramState LearnBranchingConstraint(ProgramState state, IBi
     private static ProgramState LearnBranchingConstraint<T>(ProgramState state, IBinaryOperationWrapper binary, bool falseBranch)
         where T : SymbolicConstraint
     {
+        var useOpposite = falseBranch ^ binary.OperatorKind.IsNotEquals();
         // We can fall through ?? because "constraint" and "testedSymbol" are exclusive. Symbols with the constraint will be recognized as "constraint" side.
         if ((OperandConstraint(binary.LeftOperand) ?? OperandConstraint(binary.RightOperand)) is { } constraint
-            && (OperandSymbolWithoutConstraint(binary.LeftOperand) ?? OperandSymbolWithoutConstraint(binary.RightOperand)) is { } testedSymbol)
+            && (OperandSymbolWithoutConstraint(binary.LeftOperand) ?? OperandSymbolWithoutConstraint(binary.RightOperand)) is { } testedSymbol
+            && (!testedSymbol.GetSymbolType().IsNullableBoolean() || !useOpposite))
```

Comment:
NIT: Reverse order as `!oppossite` is much more lightweight
```suggestion
            && (!useOpposite || !testedSymbol.GetSymbolType().IsNullableBoolean()))
```

Reply:
`testedSymbol.GetSymbolType().IsNullableBoolean()`

Is this part right for both `BoolConstraint` and `ObjectConstraint`? I don't think so, or do I miss something:
* `unknownNullableBool != true` -> should not learn for ObjectConstraint and BoolConstraint ✅ 
* `unknownNullableBool != null` -> should learn the opposite for ObjectConstraint "NotNull" ❓

Reply:
`unknownNullableBool != null` is the `arg != null` case on line 212.

The opposite of `NotNull` is `we have no idea => we return <null> instead of ObjectConstraint.Null`.

So the behavior makes sense.

Reply:
On line 212 arg is `int?` I was talking about the missing case `bool?`. I think for `bool?` and `!= null` we don't learn "NotNull" anymore with the change above.